### PR TITLE
Improve performance of MFA ceremony

### PR DESCRIPTION
**Pull request by rosstimothy**
To date clients attempting to access a resource first have to call `proto.AuthService/IsMFARequired` to determine if an mfa ceremony is needed for access to a resource. In an effort to reduce an extra round trip to the Auth server this can can be bundled into `proto.AuthService/GenerateUserSingleUseCerts`.

In order for RBAC to determine if mfa is required for SSH sessions the OS login of the session must be known. To accomodate this a new `SSHLogin` field was added to `proto.UserCertsRequest`.

The response to the initial request of the stream now contains a `proto.MFARequired` enum which indicates whether mfa is required, not required, or it's unknown if mfa is required. The last variant should only be returned when the `SSHLogin` field is unset in the initial request.

The `(auth.Server) isMFARequired` check was also modified for nodes to make use of `ListResources`. Instead of retrieving **all** nodes into memory and finding the matching ones, a request is made to `ListResources` with the `SearchKeywords` populated with the target from `proto.IsMFARequiredRequest_Node.Node.Node`. Care was taken to filter out any matches from labels to preserve the original matching behavior.

### CODE REVIEW

**Review by Joerger — APPROVED** [review]

**Review by Joerger — CHANGES_REQUESTED** [review]
Sorry, I didn't mean to approve yet ^

File: lib/auth/grpcserver_test.go [hunk 1]
```diff
@@ -1558,9 +1660,18 @@ func testGenerateUserSingleUseCert(ctx context.Context, t *testing.T, cl *Client
...
-	authResp := opts.authHandler(t, authChallenge.GetMFAChallenge())
+
+	challenge := authChallenge.GetMFAChallenge()
+	opts.mfaRequiredHandler(t, challenge.MFARequired)
+
+	authResp := opts.authHandler(t, challenge)
 	err = stream.Send(&proto.UserSingleUseCertsRequest{Request: &proto.UserSingleUseCertsRequest_MFAResponse{MFAResponse: authResp}})
-	require.NoError(t, err)
+	if challenge.MFARequired == proto.MFARequired_MFA_REQUIRED_NO && err != nil {
+		require.ErrorIs(t, err, io.EOF)
```
Comment (codingllama) [inline lib/auth/grpcserver_test.go:1670]:
```suggestion
		require.ErrorIs(t, err, io.EOF, "Want the server to close the stream when MFA is not required")
```

File: lib/auth/grpcserver_test.go [hunk 2]
```diff
@@ -1558,9 +1660,18 @@ func testGenerateUserSingleUseCert(ctx context.Context, t *testing.T, cl *Client
...
 	}
-	authResp := opts.authHandler(t, authChallenge.GetMFAChallenge())
+
+	challenge := authChallenge.GetMFAChallenge()
+	opts.mfaRequiredHandler(t, challenge.MFARequired)
+
+	authResp := opts.authHandler(t, challenge)
 	err = stream.Send(&proto.UserSingleUseCertsRequest{Request: &proto.UserSingleUseCertsRequest_MFAResponse{MFAResponse: authResp}})
-	require.NoError(t, err)
+	if challenge.MFARequired == proto.MFARequired_MFA_REQUIRED_NO && err != nil {
```
Comment (codingllama) [inline lib/auth/grpcserver_test.go:1669]: I'm surprised the error happens here. Based on how I've seen other streams behave, I would expect the Send to succeed and the error to only pop up in the next Recv.

File: lib/client/api.go [hunk 3]
```diff
@@ -1588,7 +1588,7 @@ func (tc *TeleportClient) ConnectToNode(ctx context.Context, clt *ClusterClient,
 	// Only return the error from connecting with mfa if the error
 	// originates from the mfa ceremony. If mfa is not required then
 	// the error from the direct connection to the node must be returned.
-	if mfaErr != nil && !errors.Is(mfaErr, MFARequiredUnknownErr{}) {
+	if mfaErr != nil && !errors.Is(mfaErr, MFARequiredUnknownErr{}) && !errors.Is(mfaErr, services.ErrSessionMFANotRequired) {
```
Comment (codingllama) [inline lib/client/api.go:1591]: Could we get this covered by tests, somehow? Is it worth the trouble?

File: lib/client/cluster_client.go [hunk 4]
```diff
@@ -121,23 +83,32 @@ func (c *ClusterClient) SessionSSHConfig(ctx context.Context, user string, targe
...
-		// In case of MFA connect to root teleport proxy instead of JumpHost to request
-		// MFA certificates.
-		c.tc.JumpHosts = nil
-		clt, err := c.tc.ConnectToCluster(ctx)
-		c.tc.JumpHosts = jumpHosts
-		if err != nil {
+	closer := func() error { return nil }
+	if target.Cluster != rootClusterName {
+		if err := c.tc.WithoutJumpHosts(func(tcNoJump *TeleportClient) error {
+			clt, err := c.tc.ConnectToCluster(ctx)
```
Comment (codingllama) [inline lib/client/cluster_client.go:89]:
```suggestion
			clusterClient, err := c.tc.ConnectToCluster(ctx)
```

(or whatever specific this is, it it's not that)

File: lib/client/cluster_client.go [hunk 5]
```diff
@@ -312,6 +284,27 @@ func performMFACeremony(ctx context.Context, clt *ClusterClient, params ReissueP
 	if mfaChal == nil {
 		return nil, trace.BadParameter("server sent a %T on GenerateUserSingleUseCerts, expected MFAChallenge", resp.Response)
 	}
+
+	switch mfaChal.MFARequired {
+	case proto.MFARequired_MFA_REQUIRED_NO:
+		return nil, trace.Wrap(services.ErrSessionMFANotRequired)
+	case proto.MFARequired_MFA_REQUIRED_UNSPECIFIED:
+		if params.MFACheck != nil && !params.MFACheck.Required {
```
Comment (codingllama) [inline lib/client/cluster_client.go:292]: In what situation would `params.MFACheck` be non-nil here? Is it trustworthy without a previous call to `IsMFARequired`?

(I genuinely don't know, so I assume a "casual" observer wouldn't either.)

File: lib/client/cluster_client.go [hunk 6]
```diff
@@ -312,6 +284,27 @@ func performMFACeremony(ctx context.Context, clt *ClusterClient, params ReissueP
...
+
+		check, err := clt.AuthClient.IsMFARequired(ctx, params.isMFARequiredRequest(clt.tc.HostLogin))
+		if err != nil {
+			return nil, trace.Wrap(MFARequiredUnknown(err))
+		}
+		if !check.Required {
+			return nil, trace.Wrap(services.ErrSessionMFANotRequired)
+		}
+
+		params.MFACheck = check
```
Comment (codingllama) [inline lib/client/cluster_client.go:304]: Why is it relevant to set this here? Should we be setting it in the other cases too?

Alternatively, could we deprecate/remove this field, or is that detrimental?

File: lib/client/cluster_client.go [hunk 7]
```diff
@@ -312,6 +284,27 @@ func performMFACeremony(ctx context.Context, clt *ClusterClient, params ReissueP
...
+	case proto.MFARequired_MFA_REQUIRED_YES:
```
Comment (codingllama) [inline lib/client/cluster_client.go:305]: Suggestion:

```suggestion
	case proto.MFARequired_MFA_REQUIRED_YES:
		// Proceed and prompt for MFA below.
```

File: integration/integration_test.go [hunk 8]
```diff
@@ -1281,25 +1281,26 @@ func testEscapeSequenceNoTrigger(t *testing.T, terminal *Terminal, sess <-chan e
 
 type localAddr struct {
 	mu   sync.Mutex
-	addr net.Addr
+	addr map[string]struct{}
```
Comment (codingllama) [inline integration/integration_test.go:1284]: Rename to plural and document key?

```suggestion
	addrs map[string]struct{} // key is the string representation of a net.Addr, eg: "localhost:8080"
```

(Correct it wrong, ofc.)

File: integration/integration_test.go [hunk 9]
```diff
@@ -1281,25 +1281,26 @@ func testEscapeSequenceNoTrigger(t *testing.T, terminal *Terminal, sess <-chan e
 
 type localAddr struct {
```
Comment (codingllama) [inline integration/integration_test.go:1282]: Should we rename the type to `localAddrs`? Should we instead be using multiple `localAddr` instances, or keeping a map of them elsewhere?

The fact that it holds multiple addresses is not something I'd expect from the present type name.

```suggestion
type localAddrs struct {
```

File: integration/integration_test.go [hunk 10]
```diff
@@ -1423,7 +1424,7 @@ func testIPPropagation(t *testing.T, suite *integrationTestSuite) {
 					return nil, trace.Wrap(err)
 				}
 
-				local.set(conn.LocalAddr())
+				local.add(conn.LocalAddr())
```
Comment (codingllama) [inline integration/integration_test.go:1427]: I don't follow what the changes here have to do with the rest of the PR.

(It looks OK by itself, I just missing how it fits.)

**Comment by codingllama** [comment]
Looks good, nice to see this one out for review.

**Review by Joerger — APPROVED** [review]

**Review by ryanclark — APPROVED** [review]

File: tool/tsh/tsh_test.go [hunk 11]
```diff
@@ -1010,206 +1024,251 @@ func TestSSHOnMultipleNodes(t *testing.T) {
...
 						RPID: "127.0.0.1",
 					},
+					RequireMFAType: types.RequireMFAType_SESSION,
 				},
 			},
-			hostLabels:      "env=dev",
+			setup:           setupChallengeSolver(successfulChallenge),
+			target:          "env=dev",
 			errAssertion:    require.Error,
+			mfaPromptCount:  0,
```
Comment (codingllama) [inline tool/tsh/tsh_test.go:1220]: Suggestion:

```suggestion
```

File: lib/client/cluster_client.go (hunk 5 above)
Comment (codingllama) [reply lib/client/cluster_client.go]: That's much clearer, thanks. I wish we had some of this background documented in the field itself.

**Review by codingllama — APPROVED** [review]

**Pull request merged**